docs/man/findbpm.1
.TH findbpm 1 "January 16th, 2003"
.SH NAME
findbpm \- average keypresses over the length of a song to find the BPM
.SH SYNOPSIS
\fBfindbpm \fIsong.ogg\fR
.SH DESCRIPTION
This script plays a song file (in Ogg Vorbis or MP3 format) and lets you
tap along on the spacebar, calculating the BPM (beats per minute) in the
song as you go. It also gives you an offset in milliseconds of when
you started tapping, compared to when the song started. You can then
use these two values to create a step file for that song.
\fI/usr/share/doc/pyddr/stepfile.txt\fR).
.SH BUGS
This is only as accurate as your sense of rhythm. :)
.SH AUTHOR
Brendan Becker <tgz@orotech.net> is the primary author of pyDDR. Joe
Wreschnig <piman@debian.org> is the maintainer of pyDDR for Debian and the
author of this manual page.
.SH SEE ALSO
\fBpyddr\fR(6)


ChangeLog
pydance ChangeLog

legend:
+ new feature added or made an existing one better
- removed a feature, or something bad happened
* fixed a bug

----- - CURRENT
0.7.5 - That's the second largest ChangeLog I've ever seen!	[2003.XX.XX]
+ New options: "Wide" to add jumps, holds on/off.
+ 5 panel (Pump It Up) support (single, versus/unison, couple, and double).
 + Support for KSF, SM, and .dance files with 5 panel steps.
 - Actual PIU pads may or may not work; none were available for testing.
+ Versus, couple, and double 6 and 8 panel modes.
+ New game type selector, so you can do e.g. Double/Endless.
+ A document on the KSF file format has been added in docs/.
+ The .dance specification has been updated with new modes and corrections.
+ A new centered display for single player modes.
+ New theming code.
 + Brings back oldschool and bryan themes.
 + Autodetect what game modes a theme will work for.
+ New 32x32 arrow set for 8/9 panel modes, by xsdg.
* Support the Thrustmaster USB mats (probably), and X-Box controllers.
 + We suspect that pydance will run on GNU/Linux on an X-Box now.
* Correctly encode/decode KSF, SM, DWI, and .dance file metadata.
 * KSF requires the Python Koco codecs package.
* DWI parsing for braindead constructs like "2!(2".
* Arrows now always draw in order (earlier under later).
* Many crashes fixed:
 * Stops + holds works again.
 * Oni lifebar works again.
 * Obscure edge case in 2 player endless mode works correctly.
* Parsing for all files is now more informative and less error-prone.
- Assist mode is broken for non-4 panel modes, we need more sounds.
- any2dance removed, it won't work anymore with this version.
- In-game menu help removed, due to inaccuracies.

0.7.4 - In My Pants						[2003.07.28]
+ 6 panel ("solo") and 8 panel support.
+ Double (1 player, both sides) support.
 - 'totaljudgings' option no longer works (no one was using it anyway...)
* Song selector UI tweaks (faster animations, nicer folder opening).
* Respace the arrows to be closer together in single/versus.
* All 9 directions can always be done on the numeric pad for player 1.
 - azerty 2nd player start key changed to '+' on the numeric pad.
* "Escape" changed from left+start to start+select.
* Some more speed optimizations for slower processors.
* Some DWI parser fixes for bad files.
* 'usepsyco' option (on by default) to try and load the Psyco JIT compiler.
- 'oldschool' and 'bryan' were removed (temporarily, maybe).

0.7.3
* Skipped, because the Linux Game Tome called 0.7.2 0.7.3 by mistake.

0.7.2 -	Dilaudid, how I love thee                               [2003.07.14]
+ Boost, brake, left, right, mirror, shuffle, and random ("supershuffle").
+ "Suck" lifebar mode, you fail if you do too good.
+ "Drop" lifebar mode, goes down at the regular speed but never up.
+ Folders in the song selector (needs > 60 songs).
* Parser bug fixes (especially for SM files).
* pyDDR has been renamed to pydance.
	
0.7.1 - OUCH MY BLADDER                                         [2003.07.01]
+ Support for COUPLE mode.
 + And battle mode.
 - any2dance doesn't always convert COUPLE from DWI to .dance correctly.
+ New player options screen (hold down start in the song selector).
+ New game options screen (hold down select).
+ "Ready? Go!" screen before the song starts.
+ New "oldschool" theme similar to earlier mixes.
* Standard round of bug fixes and tweaks.

0.7.0 - Applesauce makes the apples go yonder                   [2003.06.04]
+ More optimizations from JZig (movies this time)
+ New file formats
  + Native DWI and SM support
  + ".dance" instead of ".step" is now pyDDR's official format.
  + 'any2dance.py' converter for files.
* Judge rewrite!
  * Proper speed change handling in Judge
  * Proper miss handling in Judge

0.6.5 - Boy, did I mess that up.				[2003.05.30]
+ Major graphics optimization patch (1.5x-2x increase in fps) from JZig.
* Two player mode works again.

0.6.4 - RTFM. RTFM. RTFM. That means Read the Fucking Manual.	[2003.05.29]
+ New step file parsers
  + Beginner and hardcore (smaniac/oni/challenge) difficulty.
+ New hold arrow display
- 'mixerclock' option removed (it's now the only timing method).
- 'transcolor' option removed in favor of new 'lyriccolor' syntax.
* Bug fixes, optimizations, whippings, etc.
  + Most notably, holds, bpm changes, and stops are working much nicer now.
    - But may require reconversion with dwi2step.

0.6.3 - Reasonable Starting Times!				[2003.05.17]
+ New song loading display and algorithm (at least 4-5 times faster).
+ New song selector based roughly on DDR 5th Mix.
  + Nonstop help and playlist display.
  + 'gratuitous' option to turn off the slow extras in the song selector.
+ Endless mode.
+ Configurable oni life meter.
+ New "centered" arrow scrolling style for gameplay.
+ Smaller download size by converting wavs to oggs.
  - Requires SDL_mixer 1.2.5.
+ The timer display is controllable via the 'fpsdisplay' option.
+ New theme ('blue') and a new life display for all themes.
* DDR Extreme (8th Mix) grading and scoring algorithm.
* Split off the ddrmat.o driver from the main package.
* The constant cleanup/speedup/refactor/bug fix process.

0.6.2 - No More Ugly-Ass Buttons				[2003.04.28]
- Fullscreen switching during the game doesn't work on MacOS X.
  - Use the configuration file instead, for now.
- The pN_keys option is gone in favor of 'keyboard', and the 'sortpersist'
  option was removed.
+ 'killsongonfail' was renamed to 'autofail' and made default.
+ New game menu display.
+ Better theme detection and configuration support.
+ The 'badknees' option will remove jumps.
+ A new README that actually documents all the options!
* Once again, numerous cleanups and bugs fixed.

0.6.1 - The second one						[2003.04.09]
+ Experimental install system
  + Currently only works on Windows (and doesn't do much) and UNIX.
  + Manual pages included with pyDDR now.
  + New directory layout
+ Major code cleanups and bug fixes.

utils/findbpm.py
#!/usr/bin/env python
# This is a little utility program to find the BPM for a song.

# It works by printing out the total bpm that you've tapped in since you
# started.  This should let you average out bad values.

import getopt
import sys
import os
import pygame, pygame.font, pygame.image, pygame.mixer
from pygame.locals import *

VERSION = "0.9"

def usage():
  print "findbpm " + VERSION + " - find the bpm of a song"
  print "Usage: " + sys.argv[0] + " songfile.ogg"
  print
  print """\
Press a key on time with the beat. An average BPM (for as long as you keep
tapping) will be calculated."""

def main():
  try:
    opts, args = getopt.getopt(sys.argv[1:], "vh", ["help", "version"])
  except getopt.GetoptError:
    usage()
    sys.exit(2)
    
  if len(args) != 1:
    usage()
    sys.exit(2)
    
  pygame.init()
  screen = pygame.display.set_mode((400, 48), HWSURFACE|DOUBLEBUF)
  pygame.display.set_caption('Tap to find Start, BPM')
  screen.fill((0,0,0))
  font = pygame.font.Font(None, 48)
  text = font.render('Tap to find Start, BPM', 1, (250, 80, 80))
  textpos = text.get_rect()
  textpos.centerx = screen.get_rect().centerx
  textpos.centery = screen.get_rect().centery
  screen.blit(text, textpos)
  # show name
  pygame.display.flip()

  beats = -1
  starttime = 0
  timestr = ""
  
  pygame.mixer.music.load(args[0])
  pygame.mixer.music.play()
  beatstart = pygame.time.get_ticks()
  
  while 1:
    event = pygame.event.wait()

    if event.type == QUIT or event.type == KEYDOWN and event.key == K_ESCAPE:
      break

    elif event.type == KEYDOWN:
      beats += 1
      now = pygame.time.get_ticks()
      if beats == 0:
        starttime = now
        beatstart = starttime-beatstart
      else:
        screen.fill((0,0,0))
        timestr = '%0.6s' % (1000*60.0*beats/(now-starttime))
        beatstr = 'offset %s : bpm %s' % (beatstart, timestr)
        text = font.render(beatstr, 1, (250, 80, 80))
        textpos = text.get_rect()
        textpos.centerx = screen.get_rect().centerx
        textpos.centery = screen.get_rect().centery
        screen.blit(text, textpos)
        pygame.display.flip()

  if starttime:
    print "Started song", str(beatstart), "ms in"

  if beats > 0:
    print str(beats), 'beats over', str((now-starttime)/1000.0), 'seconds yields', timestr, 'BPM'

if __name__ == '__main__':
    main()
